### DISCUSSION TITLE
Correct way to schedule tenant commands?

### DISCUSSION THREAD

**Discussion by Muffinman** [spatie/laravel-multitenancy#474] [Q&A]
Hi,

As far as I can see it's [not documented here](https://spatie.be/docs/laravel-multitenancy/v3/advanced-usage/executing-code-for-tenants-and-landlords) what you should actually put in your crontab to run tenant schedules. I'm assuming you need both:

```bash
* * * * *  php artisan schedule:run
* * * * *  php artisan tenants:artisan schedule:run
```

We have a package service provider which is scheduling commands like this:

```php
class DearInventoryServiceProvider extends PluginServiceProvider
{
    public function boot()
    {
        parent::boot();

        $this->createSchedule();
    }

    protected function createSchedule(): void
    {
        // Schedule imports
        $this->app->booted(function () {
            $schedule = app(Schedule::class);
            $schedule->command(DearImportProductsCommand::class)->hourly();
            $schedule->command(DearImportCustomersCommand::class)->hourly();
        });
    }
}
```

`DearImportProductsCommand::class` and `DearImportCustomersCommand::class` are both `TenantAware`.

With the above setup I end up with the following schedules:

```bash
$ php artisan schedule:list

  */5 * * * *  php artisan horizon:snapshot ....................................................... Next Due: 1 minute from now
  0   * * * *  php artisan integration:dear:import-products ..................................... Next Due: 21 minutes from now
  0   * * * *  php artisan integration:dear:import-customers .................................... Next Due: 21 minutes from now
```

```bash
$ php artisan tenants:artisan schedule:list

Running command for tenant `redacted1` (id: 1)...
---------------------------------------------------------

  */5 * * * *  php artisan horizon:snapshot ..................................................... Next Due: 46 seconds from now
  0   * * * *  php artisan integration:dear:import-products ..................................... Next Due: 20 minutes from now
  0   * * * *  php artisan integration:dear:import-customers .................................... Next Due: 20 minutes from now


Running command for tenant `redacted2` (id: 2)...
---------------------------------------------------------

  */5 * * * *  php artisan horizon:snapshot ..................................................... Next Due: 46 seconds from now
  0   * * * *  php artisan integration:dear:import-products ..................................... Next Due: 20 minutes from now
  0   * * * *  php artisan integration:dear:import-customers .................................... Next Due: 20 minutes from now
```

This isn't great because:

- It will be trying to snapshot horizon multiple times
- I now need to add extra logic to check whether these custom commands are actually running in tenant context
- It's messy

I an ideal world it would only try to run the tenant commands on the tenant schedule run. Is this possible? Could it be documented?

**Comment by masterix21** [comment]
Hi @Muffinman, take a look here: 
https://laravel.com/docs/10.x/scheduling#naming-unique-jobs

Please give us feedback.

Thanks

**Comment by Muffinman** [comment]
Thanks, I'll give that a try but I'm not sure it will solve the problem in my case. I do want there to be multiple jobs, and I don't really mind which server picks them up.

What I dont want is the landlord schedule to run the Tenant specific jobs, and vice versa.

I need a way to check `Tennant::current()` where the schedule jobs are defined, but during `booting()` is to early.

Is there a documented way to run the Laravel schedule with multitenancy?
